# Image Manipulation and Data Processing

Did you know that computers see digital images as nothing but NumPy arrays?
*   A **grayscale image** (black and white) is represented as a **2D array** of pixel values. Each pixel value is a number from `0` (black) to `255` (white).
*   A **color image** is a **3D array** of shape `(height, width, 3)`. The last dimension represents the 3 color channels: Red, Green, and Blue (RGB).

When you edit an image (increase brightness, change contrast, or add filters), you are performing math operations on a NumPy array.
However, color values **must stay within the safety range of `0` and `255`**. If you increase brightness by adding `50` to a pixel that is already at `220`, the sum becomes `270`. This wraps around or gets corrupted in standard computer graphics!

To solve this, we use **`np.clip()`** (clipping). This forces all numbers to stay within a strict minimum and maximum boundary.

### Simple Explanation with an Everyday Analogy
Imagine a **dimmer switch** on a wall that controls a ceiling light bulb.
*   Turning the switch up increases the electricity to make the room brighter.
*   However, the light bulb has a physical limit: it cannot get brighter than its maximum wattage (255), and it cannot get darker than being completely off (0).
*   **Clipping** is like the physical lock on the switch that stops you from spinning the knob too far and breaking the wiring. It keeps the energy safely within the bulb's operational limits.


#### Simulating and Brightening a Grayscale Image
Let's simulate a tiny 3x3 grayscale image using random numbers and increase its brightness.


In [1]:
import numpy as np

# Let's create a simulated 3x3 grayscale image
# Pixel values represent brightness (150 to 240)
image = np.array([
    [150, 180, 200],
    [220, 160, 240],
    [190, 210, 230]
])

# Attempt 1: Add 50 brightness without clipping (DANGEROUS)
unclipped = image + 50
print("Unclipped (some values exceed 255):\n", unclipped)

# Attempt 2: Add 50 brightness and use np.clip to stay in [0, 255]
clipped = np.clip(image + 50, 0, 255)
print("Clipped (capped at 255):\n", clipped)

Unclipped (some values exceed 255):
 [[200 230 250]
 [270 210 290]
 [240 260 280]]
Clipped (capped at 255):
 [[200 230 250]
 [255 210 255]
 [240 255 255]]


#### Simulating a Color (RGB) Image Matrix
A color image has 3 color channels. Let's look at its shape.


In [2]:
import numpy as np

# A simulated 2x2 color image (RGB)
# Shape: (2 rows, 2 columns, 3 color values per pixel)
color_image = np.array([
    [[255, 0, 0],   [0, 255, 0]],    # Row 0: Red pixel, Green pixel
    [[0, 0, 255],   [255, 255, 0]]   # Row 1: Blue pixel, Yellow (Red+Green) pixel
])

print("Color Image Shape:", color_image.shape)
print("Pixel at index (0,0) (RGB values):", color_image[0, 0])

Color Image Shape: (2, 2, 3)
Pixel at index (0,0) (RGB values): [255   0   0]


### Common Mistakes Beginners Make
1.  **Forgetting that unclipped images overflow**: In computer graphics, pixel arrays are usually stored as unsigned 8-bit integers (`uint8`). If you add brightness and exceed `255` without clipping, the number loops back: `256` becomes `0` (a bright pixel suddenly turns pitch black!). Always use `np.clip()`.
2.  **Confusion over axes in images**: Remembering that in color images, axis `0` is vertical height (rows), axis `1` is horizontal width (columns), and axis `2` is color depth (RGB).


#### Exercise 1
Create a mock 4x4 grayscale image with random integers between `100` and `220`. Add a brightness bonus of `60` without using clipping, and show the result.


In [3]:
import numpy as np
# Set random seed so we get the same numbers
rng = np.random.default_rng(seed=42)
image = rng.integers(100, 221, size=(4, 4))
brightened = image + 60
print("Original:\n", image)
print("Brightened without clip:\n", brightened)

Original:
 [[110 193 179 153]
 [152 203 110 184]
 [124 111 163 218]
 [189 192 186 195]]
Brightened without clip:
 [[170 253 239 213]
 [212 263 170 244]
 [184 171 223 278]
 [249 252 246 255]]


*Explanation*: Simple scalar addition adds `60` to every element. Notice that some values (`270`, `260`, `274`) are now mathematically invalid for an image because they exceed `255`.


#### Exercise 2
Take the same original image from Exercise 1. Add the brightness bonus of `60`, but use `np.clip()` to safely limit the values between `0` and `255`.


In [4]:
import numpy as np
rng = np.random.default_rng(seed=42)
image = rng.integers(100, 221, size=(4, 4))
safe_brightened = np.clip(image + 60, 0, 255)
print("Safe Brightened:\n", safe_brightened)

Safe Brightened:
 [[170 253 239 213]
 [212 255 170 244]
 [184 171 223 255]
 [249 252 246 255]]


*Explanation*: `np.clip(..., 0, 255)` replaces any value greater than `255` with `255` and any value less than `0` with `0`. The values `270`, `260`, and `274` are safely capped at `255`.


#### Exercise 3
You have a 2x2 RGB color image:
```python
rgb_image = np.array([
    [[120, 150, 200], [255, 100, 50]],
    [[80, 200, 120],  [100, 100, 100]]
])
```
Write a Python program to completely turn off the **Red channel** (set it to `0`) for all pixels, while keeping the Green and Blue channels completely untouched.


In [5]:
import numpy as np
rgb_image = np.array([
    [[120, 150, 200], [255, 100, 50]],
    [[80, 200, 120],  [100, 100, 100]]
])

# Slicing selection: select all rows, all columns, and index 0 of the color channels
# Red is index 0, Green is 1, Blue is 2
rgb_image[:, :, 0] = 0

print("Modified RGB image:\n", rgb_image)

Modified RGB image:
 [[[  0 150 200]
  [  0 100  50]]

 [[  0 200 120]
  [  0 100 100]]]


*Explanation*: We use standard slicing to select every pixel: `[:, :]`. The third index represents the RGB channels. By specifying `0` (the Red channel), we select only the red values and assign them to `0` in a single vectorized step.
